pyteal/ast/router.py

```diff
@@ -0,0 +1,547 @@
+
+from pyteal.config import METHOD_ARG_NUM_LIMIT
+from pyteal.errors import TealInputError
+from pyteal.types import TealType
+from pyteal.compiler.compiler import compileTeal, DEFAULT_TEAL_VERSION, OptimizeOptions
+from pyteal.ir.ops import Mode
+
+from pyteal.ast import abi
+from pyteal.ast.subroutine import (
+    OutputKwArgInfo,
+    SubroutineFnWrapper,
+    ABIReturnSubroutine,
+)
+from pyteal.ast.cond import Cond
+from pyteal.ast.expr import Expr
+from pyteal.ast.app import OnComplete, EnumInt
+from pyteal.ast.int import Int
+from pyteal.ast.seq import Seq
+from pyteal.ast.methodsig import MethodSignature
+from pyteal.ast.naryexpr import And, Or
+from pyteal.ast.txn import Txn
+from pyteal.ast.return_ import Approve
+
+
+class CallConfig(Enum):
+    """
+    CallConfigs: a "bitset" alike class for more fine-grained control over
+    `call or create` for a method about an OnComplete case.
+
+    This enumeration class allows for specifying one of the four following cases:
+    - CALL
+    - CREATE
+    - ALL
+    - NEVER
+    for a method call on one on_complete case.
+    """
+
+    NEVER = 0
+    CALL = 1
+    CREATE = 2
+    ALL = 3
+
+    def __or__(self, other: object) -> "CallConfig":
+        if not isinstance(other, CallConfig):
+            raise TealInputError("OCMethodConfig must be compared with same class")
+        return CallConfig(self.value | other.value)
+
+    def __and__(self, other: object) -> "CallConfig":
+        if not isinstance(other, CallConfig):
+            raise TealInputError("OCMethodConfig must be compared with same class")
+        return CallConfig(self.value & other.value)
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, CallConfig):
+            raise TealInputError("OCMethodConfig must be compared with same class")
+        return self.value == other.value
+
+
+CallConfig.__module__ = "pyteal"
+
+
+@dataclass(frozen=True)
+class CallConfigs:
+    """
+    CallConfigs keep track of one method registration's CallConfigs for all OnComplete cases.
+    """
+
+    no_op: CallConfig = field(kw_only=True, default=CallConfig.CALL)
+    opt_in: CallConfig = field(kw_only=True, default=CallConfig.NEVER)
+    close_out: CallConfig = field(kw_only=True, default=CallConfig.NEVER)
+    clear_state: CallConfig = field(kw_only=True, default=CallConfig.NEVER)
+    update_application: CallConfig = field(kw_only=True, default=CallConfig.NEVER)
+    delete_application: CallConfig = field(kw_only=True, default=CallConfig.NEVER)
+
+    def is_never(self) -> bool:
+        return (
+            self.no_op == CallConfig.NEVER
+            and self.opt_in == CallConfig.NEVER
+            and self.close_out == CallConfig.NEVER
+            and self.clear_state == CallConfig.NEVER
+            and self.update_application == CallConfig.NEVER
+            and self.delete_application == CallConfig.NEVER
+        )
+
+    def oc_under_call_config(self, call_config: CallConfig) -> list[EnumInt]:
```

Comment:
@ahangsu Like an earlier thread, I am wondering if the highlighted method can be marked private?

I opened https://github.com/algorand/pyteal/pull/362 to propose the change.  The _optional_ PR additionally proposes a refactoring around `oc_under_call_config` usage.

Reply:
@ahangsu Thanks for bringing in #362 - resolving.